specification/eventgrid/data-plane/Microsoft.EventGrid/stable/2018-01-01/EventGrid.json

```diff
@@ -119,9 +119,19 @@
                 }
             }
         },
+        "EventGridEventData": {
```

Comment:
What is the intent with this type?  From the description it sounds like it's intended to be polymorphic?  If that's the case have you looked into using discriminated types?

Reply:
The idea behind the new EventGridEvent*Data* type is just to serve as a common base across all the first party eventdata types. More info on how I arrived at this is in approach 2 below. 

The existing EventGridEvent type contains the common properties of an event (id, eventType, eventTime etc.) + a "data" portion which is of type object. With this, there are two possible approaches:

Approach 1. [Evaluated but not used]: Specify the eventType property in EventGridEvent as the discriminator (e.g. "Microsoft.Storage.BlobCreated" -> StorageBlobCreatedEvent, "Microsoft.Storage.BlobDeleted" -> StorageBlobDeletedEvent.  Remove the "Data" property from the EventGridEvent type and have new types such as StorageBlobCreatedEvent which contain a property called Data which is of the appropriate type such as StorageBlobCreatedEventData. However, in this approach, handling for 3rd party types is not as seamless: if have a 3rd party application publishes an event with eventType "Contoso.Items.ItemReceived", they will have to have special attributes/annotations for their custom type in order to fit into this discriminator model. Further, for 3rd party events, we also support other types for the data (strings/numeric/arrays etc.), which doesn't work well with this approach. In addition, we have already GA'ed with the approach where EventGridEvent contains a Data property, so I guess it will also be a breaking change to remove it now (in order to add it to the inherited types).

Approach 2 (Current approach):  The "Data" property is part of the EventGridEvent type itself and it is of type object. The consuming code calls an EventGrid SDK function to deserialize EventGridEvents. This SDK function will internally maintain a mapping of eventType string -> eventData type and will use this mapping to look at the eventType property in the event(e.g. "Microsoft.Storage.BlobCreated") and will deserialize the Data property into the right data type such as StorageBlobCreatedEventData. The same SDK function will also support adding custom 3rd party mappings (e.g. "Contoso.Items.ItemReceived" -> typeof(ContosoItemReceivedEventData)). While creating this SDK function, I wanted to have a unit test to verify that for every system eventData type, there exists a corresponding mapping in this internal dictionary. However, without a common base class such as this newly introduced EventGridEventData, there's no way to programmatically determine which are all the eventData classes. This is the intent behind adding this new type.

Please let me know you have questions/comments.

Reply:
So if I understand correctly, the fact that some type derives from EventGridEventData indicates means that it's an eventData class correct (presumably you use reflection to make this determination)?  It seems like a very C#-specific implementation.  E.g. for Go this just generates a useless type that pollutes the package namespace (see [this](https://github.com/Azure/azure-sdk-for-go/pull/2128/files)).  I don't know how this would work in other languages.

Reply:
Yes, the fact that a type derives from EventGridEventData indicates means that it's an eventData class. For Go, if this type is not useful, can we customize the code generation to not generate it (e.g. through the --generate-empty-classes option)?  For Java, Python, and C# this would show up as a base class.

Reply:
Presumably there's a way to omit it but ideally we should have a solution that works for all languages.  How do you plan on supporting this in Go?

Reply:
My understanding from speaking to @marstr earlier this week was that this was kind of expected for Go (since the base type doesn't have any properties, nothing will be copied over). Do you or @marstr have any suggestions on if/how this can be solved in Go?

Reply:
What code will be used to programmatically determine the eventData classes, is this something your team will provide?  Do you have a sample you can share that consumes this new functionality?

Reply:
I think we can close this for now and bring it back if/when there are additional product scenarios.

Currently, it is only unit test code for a SDK function to verify that there exists a mapping for each such eventData type (for which it uses reflection to determine the count of these classes) - the idea was to catch bugs whereby somebody introduces a new eventData type but forgets to add it to the mapping dictionary.